components/director/pkg/resource/resource.go

```diff
@@ -1,77 +1,179 @@
+	// Tenant type represents tenant resource.
 	Tenant Type = "tenant"
-	// TenantIndex missing godoc
-	TenantIndex Type = "tenantIndex"
-	// Schema missing godoc
+	// TenantAccess type represents tenant access resource.
+	TenantAccess Type = "tenantAccess"
+	// Schema type represents schema resource.
 	Schema Type = "schemaMigration"
 )
 
-// SQLOperation missing godoc
+var tenantAccessTable = map[Type]string{
+	// Tables
+
+	Application: "tenant_applications",
+	Runtime:     "tenant_runtimes",
+
+	// Views
+
+	RuntimeContext:        "runtime_contexts_tenants",
+	Label:                 "labels_tenants",
+	ApplicationLabel:      "application_labels_tenants",
+	RuntimeLabel:          "runtime_labels_tenants",
+	RuntimeContextLabel:   "runtime_contexts_labels_tenants",
+	Bundle:                "bundles_tenants",
+	Package:               "packages_tenants",
+	Product:               "products_tenants",
+	Vendor:                "vendors_tenants",
+	Tombstone:             "tombstones_tenants",
+	DocFetchRequest:       "document_fetch_requests_tenants",
+	APISpecFetchRequest:   "api_specifications_fetch_requests_tenants",
+	EventSpecFetchRequest: "event_specifications_fetch_requests_tenants",
+	APISpecification:      "api_specifications_tenants",
+	EventSpecification:    "event_specifications_tenants",
+	Document:              "documents_tenants",
+	BundleInstanceAuth:    "bundle_instance_auths_tenants",
+	API:                   "api_definitions_tenants",
+	EventDefinition:       "event_api_definitions_tenants",
+	Webhook:               "webhooks_tenants",
+	AppWebhook:            "application_webhooks_tenants",
+	RuntimeWebhook:        "runtime_webhooks_tenants",
+}
+
+var parentRelation = map[Type]Type{
+	RuntimeContext:        Runtime,
+	RuntimeLabel:          Runtime,
+	RuntimeContextLabel:   RuntimeContext,
+	ApplicationLabel:      Application,
+	Bundle:                Application,
+	Package:               Application,
+	Product:               Application,
+	Vendor:                Application,
+	Tombstone:             Application,
+	DocFetchRequest:       Document,
+	APISpecFetchRequest:   APISpecification,
+	EventSpecFetchRequest: EventSpecification,
+	APISpecification:      API,
+	EventSpecification:    EventDefinition,
+	Document:              Bundle,
+	BundleInstanceAuth:    Bundle,
+	API:                   Application,
+	EventDefinition:       Application,
+	AppWebhook:            Application,
+	RuntimeWebhook:        Runtime,
+}
+
+// TenantAccessTable returns the table / view with tenant accesses of the given type.
+func (t Type) TenantAccessTable() (string, bool) {
+	tbl, ok := tenantAccessTable[t]
+	return tbl, ok
+}
+
+// Parent returns the parent type of the given type.
+func (t Type) Parent() (Type, bool) {
+	parent, ok := parentRelation[t]
+	return parent, ok
+}
+
+// TopLevelEntities is a set of entities that has a many-to-many relationship with the tenants.
+var TopLevelEntities = []Type{Application, Runtime}
```

Comment:
this mapping seems unneeded - you can derive it from the parent-relation information (if someone has no parent "he is" top level)

Reply:
There are entities which have ref specific types and generic types. For example you have ApplicationWebhook and you have ApplicationTemplateWebhook. But for some cases where you don't know which type is the parent you have generic just Webhook. Those times, the entity does not have parent, but is not top level, it is child entity however, the parent is unknown.

It should be treated as child entity, but the view that is used in those cases is aggregated for all possible parents.

One more use-case for this is when a subaccount is moved. You need to process each and every top level entity and reconcile the accesses of their m2m tables, so you need to know which of them are top level and which of them not.

Reply:
ok thats a good point. Then another idea - we could put it in the constuctor (example check func NewRepository(conv APIDefinitionConverter) *pgRepository) - there we already pass some table names to the unified repositories below, we could pass this one also. Then you can remove the